### PR TITLE
[ready] Move bernoulli into ATen

### PR DESCRIPTION
**Pull request by ssnl**
## Fixes 
+ https://github.com/pytorch/pytorch/issues/10236 : torch.bernoulli's out kwarg is broken
  fixed in moving `bernoulli_out` to ATen
+ https://github.com/pytorch/pytorch/issues/9917 : BUG torch.bernoulli(p.expand(shape)) is broken
  fixed in moving all `bernoulli` ops in ATen to use the modern apply utils methods
+ https://github.com/pytorch/pytorch/issues/10357 : torch.bernoulli inconsistent gpu/cpu results
  fixed by adding CUDA asserts

## Notable changes:

In order to use `curand_uniform4`, I made some changes to `CUDAApplyUtils.cuh`. Specifically, I introduced an optional template parameter `int step` to the `CUDA_tensor_applyN` methods, representing that we want to process `step` values at each time for each of the `N` tensors. 

The calling convention for `step = 1` (default) isn't changed. But if `step > 1`, the given lambda `op` must take in `int n` as its first argument, representing the number of valid values, because there may not be full `step` values at the boundary. E.g., here is what the `bernoulli(self, p_tensor)` call look like:
```cpp

  // The template argument `4` below indicates that we want to operate on four
  // element at each time. See NOTE [ CUDA_tensor_applyN helpers ] for details.
  at::cuda::CUDA_tensor_apply2<scalar_t, prob_t, 4>(
      ret, p,
      [seeds] __device__(
          int n, scalar_t& v1, scalar_t& v2, scalar_t& v3, scalar_t& v4,
          const prob_t& p1, const prob_t& p2, const prob_t& p3, const prob_t& p4) {
        curandStatePhilox4_32_10_t state;
        curand_init(
            seeds.first,
            blockIdx.x * blockDim.x + threadIdx.x,
            seeds.second,
            &state);
        float4 rand = curand_uniform4(&state);
        switch (n) {
          case 4: {
            assert(0 <= p4 && p4 <= 1);
            v4 = static_cast<scalar_t>(rand.w <= p4);
          }
          case 3: {
            assert(0 <= p3 && p3 <= 1);
            v3 = static_cast<scalar_t>(rand.z <= p3);
          }
          case 2: {
            assert(0 <= p2 && p2 <= 1);
            v2 = static_cast<scalar_t>(rand.y <= p2);
          }
          case 1: {
            assert(0 <= p1 && p1 <= 1);
            v1 = static_cast<scalar_t>(rand.x <= p1);
          }
        }
      }
    );
``` 

## Benchmarking

Benchmarking on `torch.rand(200, 300, 400)` 20 times, each time with 20 loops:

post patch
```
➜  ~ numactl --cpunodebind 1 --membind 1 -- taskset -c 12,13,14,15,16,17,18,19,20,21,22,23 env CUDA_LAUNCH_BLOCKING=1 python bern.py
torch.bernoulli(x)
6.841588497161865 +- 0.05413117632269859
torch.bernoulli(xc)
0.05963418632745743 +- 0.0008014909108169377
x.bernoulli_()
0.4024486541748047 +- 0.0021550932433456182
xc.bernoulli_()
0.02167394384741783 +- 2.3818030967959203e-05

```

pre-patch
```
➜  ~ numactl --cpunodebind 1 --membind 1 -- taskset -c 12,13,14,15,16,17,18,19,20,21,22,23 env CUDA_LAUNCH_BLOCKING=1 python bern.py
torch.bernoulli(x)
12.394511222839355 +- 0.0966421514749527
torch.bernoulli(xc)
0.08970972150564194 +- 0.0038722590543329716
x.bernoulli_()
1.654480218887329 +- 0.02364428900182247
xc.bernoulli_()
0.058352887630462646 +- 0.003094920190051198

```

### CODE REVIEW

File: aten/src/ATen/native/native_functions.yaml [hunk 1]
```diff
@@ -223,17 +223,22 @@
 - func: batch_norm(Tensor input, Tensor? weight, Tensor? bias, Tensor? running_mean, Tensor? running_var, bool training, double momentum, double eps, bool cudnn_enabled) -> Tensor
   variants: function
 
-- func: bernoulli(Tensor self, Tensor p, Generator* generator=nullptr) -> Tensor
```
Comment (SsnL) [inline aten/src/ATen/native/native_functions.yaml]: These two variants aren't documented or used, and they are more like tensor factory methods but not written according to the standard. So I deleted them.

File: aten/src/ATen/native/Distributions.cpp [hunk 2]
```diff
@@ -104,53 +115,143 @@ int64_t sample_poisson(double lambda, THGenerator* generator) {
...
+      vslNewStream(&stream, VSL_BRNG_MCG31, seed);
+      vslSkipAheadStream(stream, line_index_offset);
+      viRngBernoulli(VSL_RNG_METHOD_BERNOULLI_ICDF, stream, line_seg_len,
+        sample_int_ptr + line_index_offset, p);
+      vslDeleteStream(&stream);
+
+      if (contig) {
+        scalar_t *self_seg = self_ptr + line_index_offset;
+        int* tmp_seg = sample_int_ptr + line_index_offset;
+        at::vec256::convert<int, scalar_t>(tmp_seg, self_seg, line_seg_len);
```
Comment (SsnL) [inline aten/src/ATen/native/Distributions.cpp]: okay i know why this is slower. fix incoming

**Comment by fmassa** [comment]
ping @SsnL

**Comment by SsnL** [comment]
yeah I'll fix windows error

File: aten/src/ATen/core/Tensor.h [hunk 3]
```diff
@@ -441,12 +441,10 @@ struct AT_API Tensor {
   Tensor & atan_();
   Tensor baddbmm(const Tensor & batch1, const Tensor & batch2, Scalar beta=1, Scalar alpha=1) const;
   Tensor & baddbmm_(const Tensor & batch1, const Tensor & batch2, Scalar beta=1, Scalar alpha=1);
-  Tensor bernoulli(const Tensor & p, Generator * generator=nullptr) const;
```
Comment (ezyang) [inline aten/src/ATen/core/Tensor.h]: Err, was removing this intentional?

Comment (SsnL) [reply aten/src/ATen/core/Tensor.h]: Yep. It didn't quite make any sense and not used. 
What it used to do: sample `Bernoulli` according to each value in `p`. It doesn't have anything to do with `self`!!!
So it makes much more sense to just call `torch.bernoulli(p)` instead.

File: aten/src/ATen/cuda/CUDAApplyUtils.cuh [hunk 4]
```diff
@@ -126,41 +133,252 @@ void rearrangeDims(detail::TensorInfo<T1, IndexType>* aInfo,
 #define AT_APPLY_THREADS_PER_BLOCK 32 * 16
 #define AT_APPLY_BLOCKS_PER_SM 4
 
+// The `step` argument is used to support Op that operates on multiple elements
```
Comment (ezyang) [inline aten/src/ATen/cuda/CUDAApplyUtils.cuh]: Could we expand this into a longer doc about what ApplyOpX does, and what all the template arguments mean?

Comment (SsnL) [reply aten/src/ATen/cuda/CUDAApplyUtils.cuh]: See the `NOTE [ CUDA_tensor_applyN helpers ]` at line 550. Let me know if I should move it up..

File: aten/src/ATen/cuda/CUDAApplyUtils.cuh [hunk 5]
```diff
@@ -251,20 +545,228 @@ inline dim3 getApplyBlock() {
...
-  [] __device__ (scalar &a_val, const scalar &b_val) { a_val = b_val * b_val; };
+/*
+  NOTE [ CUDA_tensor_applyN helpers ]
+
+  The following CUDA_tensor_applyN (where N currently can be 1, 2, 3, or 4)
+  functions apply a pointwise operator to N tensor(s).
+
+  The calling convention is
+
+  1. The tempalte arguments should be, sequentially,
```
Comment (ezyang) [inline aten/src/ATen/cuda/CUDAApplyUtils.cuh]: nit: template

File: aten/src/ATen/cuda/CUDAApplyUtils.cuh (hunk 4 above)
Comment (ezyang) [reply aten/src/ATen/cuda/CUDAApplyUtils.cuh]: Oh, I guess the Note is below...

File: aten/src/ATen/cuda/CUDAApplyUtils.cuh [hunk 6]
```diff
@@ -126,41 +133,252 @@ void rearrangeDims(detail::TensorInfo<T1, IndexType>* aInfo,
...
+  const IndexType aOffset = sizeof...(Offsets) < n ?
+    detail::IndexToOffset<scalar, IndexType, ADims>::get(linearIndex, a) : 0;
+
+  ApplyOp1<Op, scalar, IndexType, ADims, step - 1, const IndexType, Offsets...>::apply(
+    a, op, n, linearIndex + 1, aOffsets..., aOffset
+  );
+}
+};
+
+// Specialize step=1 case. We don't need to pass in how many elements need to
```
Comment (ezyang) [inline aten/src/ATen/cuda/CUDAApplyUtils.cuh]: You say `step=1`, but I see `0` in the template parameter...

Comment (SsnL) [reply aten/src/ATen/cuda/CUDAApplyUtils.cuh]: Sorry, it's misnamed. Here the `step` is different from the `step` arg in `CUDA_tensor_applyN`. It's more like `remaining_step`. I will change them. The `step=1` comes from that we have `remaining_step=0` and only have one `typename Offset ` (vs. `typename... Offsets` below).

Comment (ezyang) [reply aten/src/ATen/cuda/CUDAApplyUtils.cuh]: Perhaps you mean, the single offset case?

Comment (SsnL) [reply aten/src/ATen/cuda/CUDAApplyUtils.cuh]: Yep, I'll clarify this.

File: aten/src/ATen/cuda/CUDAApplyUtils.cuh [hunk 7]
```diff
@@ -251,20 +545,228 @@ inline dim3 getApplyBlock() {
...
-template <typename scalar1, typename scalar2, typename Op>
-bool CUDA_tensor_apply2(at::Tensor a,
-                        at::Tensor b,
-                        Op op,
-                        TensorArgType aType = TensorArgType::ReadWrite,
-                        TensorArgType bType = TensorArgType::ReadOnly) {
+
+
+template <typename scalar, int step, typename Op>
+inline bool CUDA_tensor_apply1(at::Tensor a,
```
Comment (ezyang) [inline aten/src/ATen/cuda/CUDAApplyUtils.cuh]: I'm confused. You seem to be returning a boolean saying if the operation actually succeeds or not. But then you proceed to not check it at any of the call sites of this function. Perhaps you should replace it with an `AT_ERROR`?

File: test/test_torch.py [hunk 8]
```diff
@@ -7320,22 +7320,51 @@ def test_norm_fastpaths(self):
...
+        t.fill_(2)
+        torch.bernoulli(torch.rand_like(t, dtype=p_dtype), out=t)
+        self.assertTrue(isBinary(t))
+
+        t.fill_(2)
+        t.bernoulli_(torch.rand_like(t, dtype=p_dtype))
+        self.assertTrue(isBinary(t))
+
+    def test_bernoulli(self):
+        self._test_bernoulli(self, torch.double, 'cpu')
```
Comment (ezyang) [inline test/test_torch.py]: CUDA test?

File: test/test_torch.py [hunk 9]
```diff
@@ -7320,22 +7320,51 @@ def test_norm_fastpaths(self):
...
+
+        t.fill_(2)
+        t.bernoulli_(0.5)
         self.assertTrue(isBinary(t))
 
-        p = torch.rand(10, 10)
+        p = torch.rand(10, dtype=p_dtype, device=device).expand(10, 10)
+        t.fill_(2)
         t.bernoulli_(p)
         self.assertTrue(isBinary(t))
```
Comment (ezyang) [inline test/test_torch.py]: It makes me a little nervous that we don't have any better way of testing the bernoulli function.

File: torch/nn/init.py [hunk 10]
```diff
@@ -417,6 +417,7 @@ def deprecated_init(*args, **kwargs):
 
     See :func:`~torch.nn.init.{new_name}` for details.""".format(
         old_name=old_name, new_name=new_name)
+    deprecated_init.__name__ = old_name
```
Comment (ezyang) [inline torch/nn/init.py]: Mention this in the changelog?

**Review by ezyang — APPROVED** [review]

File: aten/src/ATen/cuda/CUDAApplyUtils.cuh (hunk 7 above)
Comment (SsnL) [reply aten/src/ATen/cuda/CUDAApplyUtils.cuh]: Very good point. I copied the return type from `CUDA_tensor_apply2` without thinking much. I'll update them all.

File: test/test_torch.py (hunk 8 above)
Comment (SsnL) [reply test/test_torch.py]: In test_cuda.py

**Pull request closed without merging**